lib/rex/proto/dns/server.rb

```diff
@@ -0,0 +1,345 @@
+  def validate_packet(packet)
+    Net::DNS::Packet.parse(
+      Resolv::DNS::Message.decode(
+        packet.respond_to?(:data) ? packet.data : packet
+      ).encode
+    )
+  end
+
+  #
+  # Process client request, handled with dispatch_request_proc if set
+  #
+  # @param cli [Rex::Socket::Tcp, Rex::Socket::Udp] Client sending the request
+  # @param data [String] raw DNS request data
+  def dispatch_request(cli, data)
+    if self.dispatch_request_proc
+      self.dispatch_request_proc.call(cli,data)
+    else
+      default_dispatch_request(cli,data)
+    end
+  end
+
+  #
+  # Default DNS request dispatcher, attempts to find
+  # response records in cache or forwards request upstream
+  #
+  # @param cli [Rex::Socket::Tcp, Rex::Socket::Udp] Client sending the request
+  # @param data [String] raw DNS request data
+  def default_dispatch_request(cli,data)
+    req = Net::DNS::Packet.parse(data)
+    forward = req.dup
+    # Find cached items, remove request from forwarded packet
+    req.question.each do |ques|
+      cached = self.cache.find(ques.qName, ques.qType.to_s)
+      if cached.empty?
+        next
+      else
+        req.answer = req.answer + cached
+        forward.question.delete(ques)
+      end
+    end
+    # Forward remaining requests, cache responses
+    if forward.question.count > 0 and @fwd_res
+      forwarded = self.fwd_res.send(validate_packet(forward))
+      req.answer = req.answer + forwarded.answer 
+      forwarded.answer.each do |ans|
+        self.cache.cache_record(ans)
+      end
+      req.header.ra = 1 # Set recursion bit
+    end
+    # Finalize answers in response
+    # Check for empty response prior to sending
+    if req.answer.size < 1
+      req.header.rCode = Net::DNS::Header::RCode.new(3)
+    end
+    req.header.qr = 1 # Set response bit
+    send_response(cli, validate_packet(req).data)
+  end
+
+protected
+  #
+  # This method monitors the listener socket for new connections and calls
+  # the +on_client_connect+ callback routine.
+  #
+  def monitor_listener
+    while true
+      rds = [self.udp_sock]
+      wds = []
+      eds = [self.udp_sock]
+
+      r,_,_ = ::IO.select(rds,wds,eds,1)
+
+      if (r != nil and r[0] == self.udp_sock)
+        buf,host,port = self.udp_sock.recvfrom(65535)
+        # Mock up a client object as a Rex Socket for sending back data
+        cli = Rex::Socket::Udp.create(
+          'PeerHost' => host,
+          'PeerPort' => port,
+          'LocalHost' => self.udp_sock.localhost,
+          'LocalPort' => self.udp_sock.localport
+        )
+        dispatch_request(cli, buf)
+      end
+    end
+  end
+
+  #
+  # Processes request coming from client
+  #
+  # @param cli [Rex::Socket::Tcp] Client sending request
+  def on_client_data(cli)
+    begin
+      data = cli.read(65535)
+
+      raise ::EOFError if not data
+      raise ::EOFError if data.empty?
+      from = [cli.peerhost, cli.peerport]
+      dispatch_request(cli, data)
+    rescue EOFError => e
```

Comment:
The main purpose of handling EOFError is to close the client? Would this be better?

```ruby
ensure close_client(cli) if cli
```